### ISSUE TITLE
Prevent manually changing Dynamic Web Module facet to >2.5 when using Java 7 runtime

### ISSUE THREAD

**Issue by chanseokoh** [GoogleCloudPlatform/google-cloud-eclipse#1965]
Previously, I think there was a facet constraint (or such a thing) that prevented having, e.g., an infeasible Java 7 + Dyn Web 3.1 facet combination. However, now it is possible to manually update the Dyn Web facet to 3.1 through the _Project Facets_ preference page.

If a project is in this situation, then uninstalling the App Engine facet and reinstalling it won't bring back the App Engine Standard Runtime library. (Probably we should also prevent installing the App Engine facet too). From this observation, it might be possible that we still have some facet constraint.

**Comment by briandealwis** [comment]
I've explored a few approaches.

In light of some of the Java 8 experiences to now (#1961, 1963, #1965, #2030, #2044, #2050), here are a few questions to drive evaluation:
  - What should be necessary to mark a project as a Java 8?  Do we want to prevent a project from being accidentally marked as Java 8?
      - Perhaps it should be driven *only* by the `<runtime>` element in `appengine-web.xml`, rather than the Project Facets?
  - is there an advantage to explicitly representing the different runtimes (jre7 vs jre8)?
     - are the Java 7 vs Java 8 environments truly interchangeable beyond the language differences?  We have a few places where we load the `appengine-web.xml` descriptor to check the `<runtime>`, and I wonder if we'll have more.

## Approach 0: _status quo_ with a facet detector ##

I thought we might be able to use the `org.eclipse.wst.common.project.facet.core.detectors` facet detectors to pull out the relevant facets.  I created a facet detector that looks for an `appengine-web.xml` and configures the facets as appropriate.

But this is only triggered by `IFacetedProjectWorkingCopy#detect()`, which is invoked by by the project context menu's _Configure > Convert to Faceted Form…_.  `#detect()` actually works pretty well, and we should use this as part of our _Convert to App Engine Standard Project_.

## Approach 1: _status quo_ with a facet validator ##

This approach adds a facet validator.  The validators are applied by the auto-build (the [`ValidationBuilder.FAMILY_VALIDATION_JOB`](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/blob/master/plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/project/ProjectUtils.java#L255) waited for in our `ProjectUtils` class).  When the validation fails, it pops up a dialog saying the build failed.  There are no problem markers or other indication of error.  There is no indication of facet conflicts in the _Project Facets_ property page.

This isn't really a great option.

## Approach 2: merge .java8 into .facets and provide two sets of constraints ##

Merge the .java8 bundle into .facets, and add two constraints to our AES facet:
```xml
   <or>
     <and><requires facet="java" version="1.7"/><requires facet="jst.web" version="2.5"/></and>
     <and><requires facet="java" version="1.8"/><requires facet="jst.web" version="[2.5-3.1]"/></and>
   </or>
```

Pros: 
- simple to implement

Cons: 
- the constraints in the _Project Facets_ property page isn't quite as clear as incompatible options are still present, and it shows a big OR with the options.
- we need to check the contents of the `appengine-web.xml` file to see if the right runtime is there. Perhaps we should just use different facet versions instead.

## Approach 3: create different versions of the `AppEngineStandardFacet` to correspond to `jre7` and `jre8` runtimes 

Our AppEngineStandardFacet (`com.google.cloud.tools.eclipse.appengine.facets.standard`) currently has a single version (_"1"_).  We would replace this with two different facet versions: one for jre7 and the other for jre8.  We'd use the `<runtime>` element to decide which one to flip.

Pros:
- separates the notion of the runtime capabilities from the supported facets (e.g., is whitelisting required?)
- no accidental enabling of Java 8
- keep the java8 work in a separate bundle, with no mixture

**Comment by chanseokoh** [comment]
I also have a few questions, which I started to raise from https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2063#issuecomment-308556495. It's about time for me to make certain things clear to me.

1. The Java facet version is fundamentally for designating the source and target compilation versions, right? For example, if a project has a Java 7 facet, you can use Java 7 specs but not Java 8 in source code, and the byte code app compiled by Eclipse will be able to run on any JVM >= 7.
1. If so, now it occurs to me that the combination of a Java 7 facet and a WTP > 2.5 facet is feasible, e.g., as hinted in https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2063#issuecomment-308556495. In this case, just having `<runtime>java8</runtime>` in `appengine-web.xml` will do, right?
1. Do/should we use the Java facet version to select which App Engine Java runtime version to use?

**Comment by briandealwis** [comment]
We went over @chanseokoh questions and:
  1. Right
  2. Yes: the AppEngine Java 8 runtime docs say existing apps should run with just a `<runtime>java8<runtime>` element: they can continue to use DWP 2.5 and Java 7
  3. We've been equating the Java version to the App Engine Java runtime version, but I think we should use an explicit App Engine Standard facet version for this instead.  Then we just have to keep the selected facet version in sync with the runtime.

We did come up with a problematic scenario that we'll have to explore: consider an app targeted towards the App Engine Java 8 runtime with `<runtime>java8<runtime>`, Java 8 facet, and DWP 3.1. The user removes the `<runtime>` element.  Both the Java and DWP facets are now wrong.  Since the user removed the `<runtime>` element, we figured that CT4E should downgrade the Java and DWP facets, and the user should see and fix any resulting compilation/validation errors.

For the converse, adding a `<runtime>java8<runtime>`, we figured we should update the Java facet to Java 8, since that's implicitly the user's intention, but leave the DWP facet unchanged.

**Comment by chanseokoh** [comment]
> CT4E should downgrade the Java (facet)

> we figured we should update the Java facet to Java 8,

You mean the App Engine Standard facet version, and not the Java facet, right? You said just above that we would not equate the Java version to the App Engine Java runtime version.

**Comment by briandealwis** [comment]
I actually mean the Java facet (`JavaFacet.VERSION_1_8`).  I think this best matches user's intent: I add `<runtime>java8</runtime>` to my class app, and now I can use Java 8'isms.  It means we have no extra steps compared to other documentation too.

(My assumption is that we want to avoid forcing users to the Project Facets page to get work done.)

**Comment by briandealwis** [comment]
Fixed by #2079 as `appengine-web.xml` now used as source of truth.

**Comment by indipirate** [comment]
If you are using javax servlet dependency then make sure you provide a correct version accordingly.
